graphql_compiler/compiler/emit_sql.py

```diff
@@ -0,0 +1,747 @@
+# Copyright 2018-present Kensho Technologies, LLC.
+"""Recursively converts a SqlNode tree to an executable SQLAlchemy query.
+
+The complexity here comes from @recurse directives (Recurse blocks). To match the semantics of
+the GraphQL compiler, recursive common table expressions (CTEs) are required. SQL backends are good
+at pushing predicates down into subqueries and CTEs, however this does not generally extend to
+recursive CTEs. This means that it is very easy to write a recursive CTE that will scan an entire
+table, even if all but a few starting points of that recursion are eventually discarded later.
+
+
+Using the query
+
+{
+    Animal {
+        name @output(out_name: "animal_name")
+             @filter(op_name: "in_collection", value: ["$names"])
+        out_Animal_LivesIn @optional {
+            name @output(out_name: "location_name")
+        }
+        out_Animal_ParentOf @recurse(depth: 2) {
+            name @output(out_name: "animal_or_descendant_name")
+        }
+    }
+}
+as an example, this is addressed with the following algorithm:
+
+1. Recursively collapse the query, treating the recursive component as a black box. For this
+example, this results in the rough SQL:
+
+SELECT
+    animal.name AS animal_name,
+    location.name AS location_name
+FROM
+    animal
+LEFT JOIN animal_livesin ON animal_livesin.animal_id = animal.animal_id
+LEFT JOIN location ON location.location_id = animal_livesin.livesin_id
+WHERE
+    animal.name IN :names
+
+2. Wrap this query as a CTE, and include any link columns in the output. A link column is the column
+that the recursive clause will later be attached to.
+
+WITH base_cte AS ( -- the actual name of the CTE is an anonymous table name
+    SELECT
+        animal.name as animal_name,
+        location.name as location_name
+        animal.animal_id as link_column -- the actual name of the column an anonymous column name
+    FROM
+        animal
+    LEFT JOIN animal_livesin ON animal_livesin.animal_id = animal.animal_id
+    LEFT JOIN location ON location.location_id = animal_livesin.livesin_id
+    WHERE
+        animal.name IN :names
+)
+
+3. Construct the recursive clause. Here we only recurse on the columns necessary to JOIN before
+and after the recursion, output columns are not carried along. The recursion is joined to the
+CTE of the base query, ensuring that the recursion only starts at the required starting points,
+no more.
+
+Also worth noting with the recursive clause is the __depth_internal_name, which keeps track of
+recursion depth per the compiler's semantics, and __path_internal_name, which is used for cycle
+detection. Simply using UNION here to join the anchor query to the recursive query is insufficient
```

Comment:
Not sure I understand what the problem with cycles is? Shouldn't we output cycles if they exist?

Reply:
My motivation was that without cycle detection, we will repeat a lot of the same traversals that we already have sometime previously, which I was avoiding to save the DB engine some work. This was mainly motivated by performance on some real world datasets I was experimenting with.

This is another question I have with semantics then. Let's say we're repeatedly traversing an edge with cycle A -> B -> C -> A, where the types all work out and we have a very high recursion depth. Do we output just the rows:
```
..., A
..., B
..., C
```
and stop with that traversal because the next row would be `... A`, starting a traversal we've already traversed, or do we continue to repeat, and continue to spit out 
```
..., A
..., B
..., C
..., A
..., B
..., C
...
```
for whatever our recursion depth limit allows?

Reply:
Semantically, it would be the latter.

One of the reasons for this is because there have been many requests for implementing a meta-field that allows outputting the recursion depth (which we haven't implemented yet because of lack of support for this in OrientDB 2). If and when we implement this feature, one should be able to query for things like "all the cycles and cycle lengths that start and end at A, to max depth 7".

Unfortunately, recursion can definitely get quite expensive, even on the other backends. This is something the compiler can help with, when we get a chance to integrate query size estimation and automatic query paging, but in the end we have to have some amount of trust that the user knows what they are asking for. Even without cycles, the database can have very deep trees and the user can set a recursion depth that is way too high -- and the resulting query will be expensive no matter what.

I'd be open to allowing distinct-only recursion in the future as an addition, if there is a good use case for it and ideally if we are able to figure out a way to emit appropriate OrientDB code for it.

Reply:
That makes sense, I'll update accordingly.

The SQL backend can output the depth relatively easily if we ever support that at the compiler level. 

I've also thought it would be very useful to get the path taken in a recursion. That would allow you to treat results in a BFS vs DFS manner on the client side if desired, depending on how the path output is used for sorting. As another future thing if it can be worked in OrientDB.

That sounds good WRT distinct recursions, it is very costly as you point out, and giving users ways to address that is good future work.